_episodes/16-parallel.md

```diff
@@ -142,4 +613,66 @@ output file, and examine it. Is it what you expected?
 
 * How good is the value for &#960;?
 * How much memory did it need?
-* How much of that memory was used on each node?
+* How long did the job take to run?
+
+Modify the job script to increase both the number of samples and the amount
+of memory requested (perhaps by a factor of 2, then by a factor of 10),
+and resubmit the job each time.
+You can also increase the number of CPUs.
+
+* How good is the value for &#960;?
+* How much memory did it need?
+* How long did the job take to run?
+
+## How Much Does MPI Improve Performance?
+
+In theory, by dividing up the &#960; calculations among *n* MPI processes,
+we should see run times reduce by a factor of *n*.
+In practice, some time is required to start the additional MPI processes,
+for the MPI processes to communicate and coordinate, and some types of
+calculations may only be able to run effectively on a single CPU.
+
+Additionally, if the MPI processes operate on different physical CPUs
+in the computer, or across multiple compute nodes, additional time is
+required for communication compared to all processes operating on a
+single CPU.
+
+[Amdahl's Law](https://en.wikipedia.org/wiki/Amdahl's_law) is one way of
+predicting improvements in execution time for a fixed parallel workload.
+If a workload needs 20 hours to complete on a single core,
+and one hour of that time is spent on tasks that cannot be parallelized,
+only the remaining 19 hours could be parallelized.
+Even if an infinite number of cores were used for the parallel parts of
+the workload, the total run time cannot be less than one hour.
+
+In practice, it's common to evaluate the parallelism of an MPI program by
+* running the program across a range of CPU counts,
+* recording the execution time on each run,
+* comparing each execution time to the time when using a single CPU.
+
+The speedup factor *S* is calcuated as the single-CPU execution time divided
+by the multi-CPU execution time.
+For a laptop with 8 cores, the graph of speedup factor versus number of cores
```

Comment:
The specifics of this can get tripped up in *cores* vs. *threads*: there'll be a severe drop at the `cores+1` point, rising back up to `cores` performance at `threads`.

Reply:
Maybe I adjust the graphs to read "number of ranks" on the horizontal axis instead of "number of cores" or threads. If ranks<=cores, should get consistent speedup. If ranks>cores, diminishing returns. If ranks>threads, should be no improvement at all.